c7n/actions.py

```diff
@@ -378,6 +390,67 @@ def process(self, resources, event=None):
             params['Payload'] = utils.dumps(payload)
             result = client.invoke(**params)
             result['Payload'] = result['Payload'].read()
+            if isinstance(result['Payload'], bytes):
+                result['Payload'] = result['Payload'].decode()
+            results.append(result)
+        return results
+
+
+class MethodInvoke(EventAction):
+    """ Invoke an arbitrary method
+
+    serialized invocation parameters
+
+     - resources / collection of resources
+     - policy / policy that is invoke the function
+     - action / action that is invoking the function
+     - version / version of custodian invoking the lambda
+
+    Example::
+
+     - type: invoke-method
+       module: my_module
+       class: actions.action_class
+       method: my_method
+    """
+
+    schema = {
+        'type': 'object',
+        'properties': {
+            'type': {'enum': ['invoke-method']},
+            'module': {'type': 'string'},
+            'class': {'type': 'string'},
+            'method': {'type': 'string'},
+            'qualifier': {'type': 'string'},
+            'batch_size': {'type': 'integer'}
+        }
+    }
+
+    def __init__(self, data=None, manager=None, log_dir=None):
+        super(MethodInvoke, self).__init__(data, manager, log_dir)
+        self.requires_permissions = False
+
+    def get_permissions(self):
+        return ()
+
+    def process(self, resources, event=None):
+        importlib.import_module(self.data['module'])
+        kwargs = dict(FunctionName=self.data['class'])
+        if self.data.get('qualifier'):
+            kwargs['Qualifier'] = self.data['Qualifier']
+
+        payload = {
+            'version': VERSION,
+            'event': event,
+            'action': self.data,
+            'policy': self.manager.data}
+
+        results = []
+        for resource_set in utils.chunks(resources, self.data.get('batch_size', 250)):
+            payload['resources'] = utils.dumps(resource_set)
+            a_class = getattr(sys.modules[self.data['module']], self.data['class'])
```

Comment:
its actually preferable to do an `__import__` or use importlib, sys.modules lookups are dependent on import ordering, also the default value here isn't valid, as you'll get a failure on the next line trying to instantiate a string. ideally you'd do this lookup in validate as well.

Reply:
Good feedback. I guess this would work as an action plugin. Could you point me to where I can see the mechanism where would deploy a plugin via setup tools?

Reply:
The azure and gcp providers in tools directory are setup that way. its basically just a bit of metadata in setup.py (see link below), when the entry point is invoked, your code will be imported and you can register any filters, actions, registries of interest. ie say you wanted to register an action on all aws resources in your plugin, it would just be. 

```
from c7n.resources.aws import AWS

def entrypoint():
   for name, resource in AWS.resources.items():
       resource.action_registry.register('custom-integration', CustomFoo)
```

https://github.com/capitalone/cloud-custodian/blob/master/tools/c7n_gcp/setup.py#L31

when custodian runs its loads up that entrypoint across all installed packages accessible to the interpreter.

All that said we encourage users to contribute filters/actions to the opensource repo, so that they can benefit other users, and be maintained along side the rest of the code base.

Reply:
@kapilt That worked perfectly, thanks! *However*, I'd like to make one more plea for the concept of having a resource-wide generic action for invoking local methods. 

The primary value here is:

* We can write complex remediation code and run it local to the custodian runner task in Fargate without having to manage the overhead of an additional lambda

* We can write methods that can inspect the returned data from the filters for local debugging. For example, we wrote a custom `invoke-method` called `print_payload` which allows us to write a complex filter and run this action to output the actual json returned in the payload from the Cloud Custodian filter, so we can see what data is available to work with.

* We can test using `invoke-method` and when we've debugged them and sufficiently tested them, convert them to `invoke-lambda` if we want.

* We've built a template framework for policies and having local methods really speeds up the development time by not having to rely on lambdas for remediation.

This isn't a Fargate vs. Lambda or local vs. Lambda . . . this is simply just another option to allow developers to debug and develop policies more quickly. I appreciate your time here. This is a killer codebase to work with . . .